### PR TITLE
[Feat][MoE] Add moe_fused_topk kernel and op.

### PR DESCRIPTION
**Pull request by michaelwithu**
## Summary

Implements `FusedTopKKernel` and `FusedTopKOp` for MoE top-k routing, closes #XXX.

Single kernel fusing scoring (softmax or sigmoid) and top-k selection — no global memory
roundtrip for intermediate scores.

- **One warp per token**: lane `l` owns experts `{l, l+32, l+64, ...}` in local registers —
  no shared memory, no inter-warp communication
- **Softmax**: 2-pass warp `shfl_xor` all-reduce (max then sum), `exp` in-place;
  normalize deferred to output write (argmax ordering is identical for `exp(x−max)` and
  `softmax(x)`, saving one full pass over E elements)
- **Sigmoid**: element-wise in-register, no reduction needed
- **K-pass argmax**: per-pass `shfl_xor` `(val, idx)` all-reduce with deterministic
  tie-breaking (prefer lower expert index on equal scores) to prevent split-brain masking
  when bf16 inputs produce identical f32 scores across two experts
- **Zero `__syncthreads()`** across all paths (vs 22 syncs for softmax in prior
  block-reduce designs, vs ~29 syncs in vLLM's CUB path)
- Supports: `softmax` (Qwen3/Qwen2/Mixtral), `sigmoid` (DeepSeek-V3/GLM-4),
  `renormalize=True/False`, bf16/fp16/fp32 input

**Inputs:**
- `gating_output [T, E]` — router logits (bf16/fp16/fp32)

**Outputs:**
- `topk_weights [T, K]` float32 — routing weights (optionally renormalized)
- `topk_ids     [T, K]` int32   — selected expert indices

## Benchmark (H200, bf16 input, top_k=8)

| num_tokens | num_experts | scoring_func | TileOPs (ms) | TileOPs (TFLOPS) | TileOPs (TB/s) | vLLM (ms) | vLLM (TB/s) | PyTorch (ms) | PyTorch (TB/s) | vs vLLM | vs PyTorch |
|---|---|---|---|---|---|---|---|---|---|---|---|
| 512  | 128 | softmax | 0.01 | 0.13 | 0.02 | 0.01 | 0.02 | 0.02 | 0.01 | **1.0×** | **2×** ↑ |
| 2048 | 128 | softmax | 0.01 | 0.49 | 0.07 | 0.01 | 0.07 | 0.03 | 0.02 | **1.0×** | **3×** ↑ |
| 4096 | 128 | softmax | 0.01 | 0.79 | 0.11 | 0.01 | 0.10 | 0.05 | 0.03 | **1.1×** ↑ | **4×** ↑ |
| 512  | 256 | softmax | 0.01 | 0.16 | 0.02 | 0.01 | 0.03 | 0.03 | 0.01 | **1.0×** | **3×** ↑ |
| 2048 | 256 | softmax | 0.02 | 0.59 | 0.07 | 0.02 | 0.07 | 0.05 | 0.02 | **1.0×** | **3×** ↑ |
| 4096 | 256 | softmax | 0.02 | 0.96 | 0.12 | 0.02 | 0.11 | 0.08 | 0.03 | **1.1×** ↑ | **5×** ↑ |
| 512  | 256 | sigmoid | 0.01 | 0.16 | 0.02 | 0.01 | 0.03 | 0.03 | 0.01 | **1.0×** | **3×** ↑ |
| 2048 | 256 | sigmoid | 0.02 | 0.59 | 0.07 | 0.02 | 0.07 | 0.05 | 0.02 | **1.0×** | **3×** ↑ |
| 4096 | 256 | sigmoid | 0.02 | 0.98 | 0.12 | 0.02 | 0.11 | 0.09 | 0.03 | **1.1×** ↑ | **4×** ↑ |

TileOPs matches or beats vLLM on all configs (10–13% faster at T=4096), and is **2–5× faster**
than PyTorch reference. Latency resolution is 0.01ms; TB/s reflects the true throughput gap.

## Tests
pytest tests/ops/test_moe_fused_topk.py -vvs  # 13/13 passed

Covers: softmax/sigmoid, renormalize=True/False, bf16/fp16, top_k=1/8,
various (T, E, K) shapes including E not divisible by 32 (padding path).

### CODE REVIEW

File: tileops/kernels/moe/fused_topk.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+                                other_idx,
+                                T.if_then_else(
+                                    other_val == l_best_val,
+                                    T.if_then_else(
+                                        other_idx < l_best_idx, other_idx, l_best_idx
+                                    ),
+                                    l_best_idx,
+                                ),
+                            )
+                            l_best_val = T.max(l_best_val, other_val)
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tileops/kernels/moe/fused_topk.py:200]: Good point. Refactoring into separate boolean conditions isn't feasible here since T.if_then_else is a TileLang IR node (not a Python conditional) — any Python-level bool would be evaluated at compile time rather than emitted as a GPU branch. Added a detailed comment explaining the structure: this implements a lexicographic comparison on (-val, idx) where higher value wins and lower expert index breaks ties, with each branch annotated explicitly. Updated in https://github.com/tile-ai/TileOPs/commit/ec10599a596c6862c768e5061353a27ec1234888.
